src/Microsoft.Diagnostics.Monitoring.StartupHook/Exceptions/Eventing/ExceptionsEventSource.cs

```diff
@@ -167,13 +186,49 @@ private unsafe void WriteEventCore(int eventId, Span<EventData> data)
             data.Size = sizeof(T);
         }
 
+        [NonEvent]
+        private static unsafe void SetValue(ref EventData data, in Span<byte> value)
+        {
+            data.DataPointer = (nint)Unsafe.AsPointer(ref value.GetPinnableReference());
+            data.Size = value.Length;
+        }
+
         [NonEvent]
         private static void SetValue(ref EventData data, in PinnedData value)
         {
             data.DataPointer = value.Address;
             data.Size = value.Size;
         }
 
+        private static int GetArrayDataSize<T>(T[] data) where T : unmanaged
+        {
+            // Arrays are written with a length prefix + the data as bytes
+            return ArrayLengthFieldSize + Unsafe.SizeOf<T>() * data.Length;
+        }
+
+        private static unsafe void FillArrayData<T>(Span<byte> target, T[] source) where T : unmanaged
+        {
+#if DEBUG
+            if (target.Length != GetArrayDataSize(source))
```

Comment:
Just for my own understanding - when might this scenario occur? My initial thought was if the `stackalloc` didn't allocate enough space the `target.Length` could be too small, but my understanding is that if we tried to allocate more than what's available we would get a `StackOverflowException` (and we'd never get here). I haven't worked with `stackalloc` before so I figured I'd clarify.

Reply:
Dev authoring error. If you try to pass a Span that wasn't sized to what was expected. Should never be an issue at runtime.